datasimulator/graph.py

```diff
@@ -201,55 +201,54 @@ def construct_graph_edges(self):
-            node(Node): current node object
-
-        Outputs:
-            list: list of submission order
-
+        From the specified `end_node`, step through the graph from bottom to top to generate the minimal
+        submission order to that node.
         """
-        submission_order = [node]
-        if cmc_node:
-            submission_order.append(cmc_node)
-        index = 0
-
-        project_node = None
-        while index < len(submission_order):
-            cur_node = submission_order[index]
-            index += 1
-            if not cur_node:
-                continue
-            for linked_node_dict in cur_node.required_links:
-                if linked_node_dict["node"].name == "project":
-                    project_node = linked_node_dict["node"]
-                    continue
-                if linked_node_dict["node"] in submission_order:
-                    # reorder to place the node at the beginning. eg. if this node is parent to 2 other
-                    # nodes, we need it to be submitted before the 2 others, and not in-between.
-                    submission_order.remove(linked_node_dict["node"])
-                submission_order.append(linked_node_dict["node"])
-
-        if project_node:  # project should always be submitted first
-            submission_order.append(project_node)
+        to_visit = [end_node]
+        submission_order = [end_node]
+        while to_visit:
+            node = to_visit.pop()
+            for parent_node_dict in node.required_links:
+                parent_node = parent_node_dict["node"]
+                if parent_node in submission_order:
+                    submission_order.remove(parent_node)
+                submission_order.append(parent_node)
+                if parent_node not in to_visit:
+                    to_visit.append(parent_node)
 
         submission_order.reverse()
 
+        # if specified, make sure that `core_metadata_collection` is in the submission order
+        if cmc_node and cmc_node not in submission_order:
+            # insert cmc node right after "project"
+            submission_order.insert(1, cmc_node)
+
         return submission_order
 
     def generate_submission_order(self):
         """
-        Generate submission order for the graph
+        Step through the graph from top to bottom to generate the submission order from the `project` node
+        to all leaf nodes.
         """
-        submission_order = []
+        # populate the nodes' `child_nodes` lists
         for node in self.nodes:
-            if node not in submission_order:
-                for item in self.generate_submission_order_path_to_node(node):
-                    if item not in submission_order:
-                        submission_order.append(item)
+            if node.name == "project":
+                proj_node = node
+            for link in node.required_links:
+                link["node"].child_nodes.append(node)
+
+        to_visit = [proj_node]
```

Comment:
if this is the same logic as in `generate_submission_order_path_to_node`, can you break it out to a utility function both call?

Reply:
almost - one of them is top to bottom, the other bottom to top. I can refactor.